cat/spirv.cat

```diff
@@ -0,0 +1,127 @@
+// loc order, relates memory accesses to the same location
+let locord = loc &
+             (  (hb & int & vloc) |
+                ([R \ PRIV]; hb; [R | W \ PRIV]) | // RaR, WaR (non-private)
+                ([R]; ssw+; [R | W]) | // RaR, WaR (any)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsg; (hb & ssg);                         [W \ PRIV])) |                                 // WaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsg; (hb & ssg); vissg; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avwg; (hb & swg);                         [W \ PRIV])) |                                 // WaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avwg; (hb & swg); viswg; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via workgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avqf; (hb & sqf);                         [W \ PRIV])) |                                 // WaW (via queue family instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avqf; (hb & sqf); visqf; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via queue family instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsh; (hb);                               [W \ PRIV])) |                                 // WaW (via shader domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsh; (hb);       vissh; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via shader domain)
+                (       ([W];        (hb & avvisinc);  avdv; (hb);                               [W])) |                                        // WaW (via device domain)
+                (       ([W];        (hb & avvisinc);  avdv; (hb);       visdv; (hb & avvisinc); [R]; (po? & avvisinc); [R \ PRIV]))            // RaW (via device domain)
+             )
+
+let w-locord = [W]; locord
+
+let visto = (w-locord \ (w-locord; w-locord)); [R] // visible to = location ordered W->R with no intervening write (W->W->R)
+(*******************)
+
+
+(****************)
+(* restrictions *)
+(****************)
+// rai[asmo] = ((A&W) -> (A&W)) & mutordatom
+empty ([A&W]; mutordatom; [A&W]) \ (asmo | asmo^-1) as coherence
+
+(***********************)
+(* Memory Model Axioms *)
+(***********************)
+// consistency: locord, rf, fr, asmo must not form cycles
+acyclic (locord | rf | fr | asmo) as consistency-cycle
+// consistency: non-atomic cannot read-from a value that is shadowed by another write
+empty (rf; ([R \ ATOM])) & (([W]; locord); ([W]; locord)+) as consistency-rf
+
+let read-write = (R \ W) * (R \ W) // exclude RMWs
+empty (((locord | locord^-1) | mutordatom \ read-write) \ (loc \ id \ read-write)) as locord-complete
+
+(* data_races *)
+let race-pair = W * W | W * R | R * W
```

Comment:
The name is confusing, any pair of these is not necessary a racy one. I would simply use `let dr = loc & (W * W | W * R | R * W \ mutordatom \ id \ (locord | locord^-1))`

Reply:
merged into dr relation

Reply:
Please add a parenthesis wrapping the old race-pair (I also forgot it in my previous comment; might not be needed but better safe than sorry)

Reply:
Wrapped